src/porepy/models/compositional_flow.py

```diff
@@ -0,0 +1,2070 @@
+
+import logging
+import time
+from functools import partial
+from typing import Callable, Optional, Sequence, cast
+
+import numpy as np
+
+import porepy as pp
+import porepy.compositional as ppc
+
+from .protocol import CompositionalFlowModelProtocol
+
+logger = logging.getLogger(__name__)
+
+
+def update_phase_properties(
+    grid: pp.Grid,
+    phase: pp.Phase,
+    props: ppc.PhaseProperties,
+    depth: int,
+    update_derivatives: bool = True,
+    use_extended_derivatives: bool = False,
+) -> None:
+    """Helper method to update the phase properties and its derivatives.
+
+    This method is intended for a grid-local update of properties and their derivatives,
+    using the methods of the
+    :class:`~porepy.numerics.ad.surrogate_operator.SurrogateFactory`.
+
+    Parameters:
+        grid: A grid in the md-domain.
+        phase: Any phase.
+        props: A phase property structure containing the new values to be set.
+        depth: An integer used to shift existing iterate values backwards.
+        update_derivatives: ``default=True``
+
+            If True, updates also the derivative values.
+        use_extended_derivatives: ``default=False``
+
+            If True, and if ``update_derivatives==True``, uses the extend
+            derivatives of the ``state``.
+
+            To be used in the the CFLE setting with the unified equilibrium formulation.
+
+    """
+    if isinstance(phase.density, pp.ad.SurrogateFactory):
+        phase.density.progress_iterate_values_on_grid(props.rho, grid, depth=depth)
+        if update_derivatives:
+            phase.density.set_derivatives_on_grid(
+                props.drho_ext if use_extended_derivatives else props.drho, grid
+            )
+    if isinstance(phase.specific_enthalpy, pp.ad.SurrogateFactory):
+        phase.specific_enthalpy.progress_iterate_values_on_grid(
+            props.h, grid, depth=depth
+        )
+        if update_derivatives:
+            phase.specific_enthalpy.set_derivatives_on_grid(
+                props.dh_ext if use_extended_derivatives else props.dh, grid
+            )
+    if isinstance(phase.viscosity, pp.ad.SurrogateFactory):
+        phase.viscosity.progress_iterate_values_on_grid(props.mu, grid, depth=depth)
+        if update_derivatives:
+            phase.viscosity.set_derivatives_on_grid(
+                props.dmu_ext if use_extended_derivatives else props.dmu, grid
+            )
+    if isinstance(phase.thermal_conductivity, pp.ad.SurrogateFactory):
+        phase.thermal_conductivity.progress_iterate_values_on_grid(
+            props.kappa, grid, depth=depth
+        )
+        if update_derivatives:
+            phase.thermal_conductivity.set_derivatives_on_grid(
+                props.dkappa_ext if use_extended_derivatives else props.dkappa, grid
+            )
+
+
+# region general PDEs used in the (fractional) multiphase, multicomponent flow.
+
+
+class DiffusiveTotalMassBalanceEquations(pp.BalanceEquation):
```

Comment:
You are right. I changed the class and it inherits now from the other version using upwinding.

Flux and source term are now the only relevant changes, which use solely the Darcy flux and respective interface fluxes.

I resolved all your remarks regarding this since they are now obsolete. Please review this class again.

Reply:
Summary of conversation: We will make the change even more atomic, by overriding interface_darcy_flux (if I remember correctly) and well_flux, instead of the flux and source terms, as is currently the state.

Reply:
`interface_fluid_flux` and `well_fluid_flux` are now overridden and point simply to the respective variable which are now massic in this setup (consistent with the Darcy flux being also massic here)